Theory assessment/Theory_assessment_Software_Engineering.md

```diff
@@ -0,0 +1,150 @@
+  - They can see later the result if voting is ended.
+  - Getting notification if results are out. 
+- MAIN CONSIDERATIONS:
+  - How to make sure that person votes only one time. 
+  - Can person change their vote until the end, just in case they choose wrong person.
+  - If there are a lot of people/similar names in list, then how to make system easy for voting, so there would not be confusion. 
+- BIGGEST/COMMON PROBLEMS:
+    - Confusion with names or people, if there is a lot of employees.
+    - People wanting to change their votes.
+- COMPONENTS/TOOLS:
+    - SQL for holding data and e-mails - who have voted.
+    - PYTHON for back-end logic.
+    - HTML and CSS for front-end.
+    - API for e-mail notifications.
+
+### 5. What's the software development cycle?
+A: Software development cycle known as SDLC, is needed to deliver expected goals to client. It has 6 steps:
+1. Planning: making plan(s) how to achieve the goal and requirements analysis.
+2. Analysis: deciding tool/methods that project is made and how these fill requirements.
+3. Design: planning and setting up the architecture, so design would have a suitable logic.
+4. Development: coding and making the goal product.
+5. Integration and testing: putting all the work pieces together and testing + documentation(architecture, technical, user).
+6. Maintenance: deploying, improving and maintaining the product.
+
+### 6. What's the difference between agile and waterfall?
+A: Waterfall is more phase-to-phase (linear) and strict approach to development. Waterfall is great for projects that have specific timeline and very concrete requirements, since everything is organized before work starts. Everything is made simple and project can be done without any hurdles. Cons of waterfall are that it is not very flexible, because if changes are made or something goes wrong, there is a chance to have to start again. It also takes time and cost can be big, if there is something that needs to be done fast. Agile is more flexible, there are not many phases and requires more testing, which leads to better customer satisfaction. Projects are divided into smaller pieces and prioritized. Cons of Agile are that there should be clear picture of the goal, team commitment and tight cooperation for project to succeed.
+ ![image](agilevswaterfall.jpg)
+
+### 7. What is a reduced function used for?
+A: Reduce function takes list and function of two arguments.
+Given functions is cumulatively applied to all list elements. 
+Reduce returns cumulated result(single value).
+```
+import functools 
+
+list= ['Code', 'First', 'Girls']
+full_name= functools.reduce(lambda x,y: x+y,list)
+print(full_name)
+
+Result: CodeFirstGirls
+```
+### 8. How does merge sort work?
+A: Merge sort is algorithm that takes array and divides it smaller and smaller sub-pieces until the dividing is impossible. Then it sorts the smallest pieces, compares them, sorts them and merges them into a list.   
```

Comment:
8/10 You could add more details on what the problem-solving technique being used in merge sort such as divide and conquer, recursive.